pkg/client/userd/trafficmgr/intercept.go

```diff
@@ -333,21 +335,110 @@ func (s *session) ensureNoInterceptConflict(ir *rpc.CreateInterceptRequest) *rpc
+		targetPort = spec.ContainerPort
+	}
+	ports := make([]agentconfig.PortAndProto, 0, len(spec.LocalPorts)+len(spec.PodPorts)+1)
+	ports = append(ports, agentconfig.PortAndProto{
+		Port:  uint16(targetPort),
+		Proto: core.Protocol(spec.Protocol),
+	})
+	for _, lp := range spec.LocalPorts {
+		pp, _ := agentconfig.NewPortAndProto(lp)
+		ports = append(ports, pp)
+	}
+	for _, ps := range spec.PodPorts {
+		pm := agentconfig.PortMapping(ps)
+		ports = append(ports, pm.To())
+	}
+	return ports
+}
+
+// ensureUniqueLocalPorts returns the sum of all local ports that the intercept will forward to, and all
+// local ports that the client will forward from. Also ensures that there are no conflicts among those ports.
+// The cluster-side of the port mappings are not checked here because we rely on the PrepareIntercept
+// call to already have done that.
+func ensureUniqueLocalPorts(spec *manager.InterceptSpec, pi *manager.PreparedIntercept) (map[agentconfig.PortAndProto]struct{}, error) {
+	targetPort := spec.TargetPort
+	if targetPort == 0 {
+		targetPort = pi.ContainerPort
+	}
+
+	ports := make(map[agentconfig.PortAndProto]struct{}, len(spec.LocalPorts)+len(spec.PodPorts)+1)
+	ports[agentconfig.PortAndProto{
+		Port:  uint16(targetPort),
+		Proto: core.Protocol(pi.Protocol),
+	}] = struct{}{}
+
+	for _, lp := range spec.LocalPorts {
+		pp, err := agentconfig.NewPortAndProto(lp)
+		if err != nil {
+			return nil, err
+		}
+		if _, ok := ports[pp]; ok {
+			return nil, fmt.Errorf("multiple use of port %s on %s", pp, spec.TargetHost)
+		}
+		ports[pp] = struct{}{}
+	}
+	for _, ps := range pi.PodPorts {
+		pm := agentconfig.PortMapping(ps)
+		if err := pm.Validate(); err != nil {
+			return nil, err
+		}
+		pp := pm.To()
+		if _, ok := ports[pp]; ok {
+			return nil, fmt.Errorf("multiple use of port %s on %s", pp, spec.TargetHost)
+		}
+		ports[pp] = struct{}{}
+	}
+	return ports, nil
+}
+
+func (s *session) ensureNoPortConflict(spec *manager.InterceptSpec, ir *manager.PreparedIntercept) *rpc.InterceptResult {
+	ports, err := ensureUniqueLocalPorts(spec, ir)
+	if err != nil {
+		return InterceptError(common.InterceptError_TRAFFIC_MANAGER_ERROR, errcat.User.New(err))
+	}
+
+	s.currentInterceptsLock.Lock()
+	defer s.currentInterceptsLock.Unlock()
+	for _, ci := range s.currentIntercepts {
+		ciSpec := ci.Spec
+		for _, blp := range allBusyLocalPorts(ciSpec) {
+			if _, ok := ports[blp]; ok {
+				return &rpc.InterceptResult{
+					Error:         common.InterceptError_LOCAL_TARGET_IN_USE,
+					ErrorText:     fmt.Sprintf("Port %s is already in use by intercept %s", net.JoinHostPort(ciSpec.TargetHost, blp.String()), ciSpec.Name),
+					ErrorCategory: int32(errcat.User),
+				}
 			}
 		}
 	}
 	return nil
 }
 
+func (s *session) compareFinalizedManagerVersion(major, minor, patch uint64) int {
+	mv := s.managerVersion
+	n := mv.Major - major
+	if n == 0 {
+		if n = mv.Minor - minor; n == 0 {
+			n = mv.Patch - patch
+		}
+	}
+	return int(n)
+}
+
```

Comment:
It might be better to turn it into `requireManagerVersion` and return the boolean value, or use `semver.Version` instead to make it clean.

```suggestion
func (s *session) satisfiesManagerVersion(r semver.Range) bool {
	return r(s.managerVersion)
}
```

and usage

```go
session.satisfiesManagerVersion(semver.MustParseRange(">=2.22.0"))
session.satisfiesManagerVersion(semver.MustParseRange("2.22.0"))
```
etc.


Alternatively, the function could return the severity as well, so major/minor/patch to be more reliable.

Some checks like the one in `ensureWatchers` may yield unexpected results if we have another patch release of traffic-manager 2.20.x (unlikely to happen, but you know what I mean).
As an example, we'll always have the result of `1` if the the traffic-manager version is `2.x.1` while inputs are `2.x.0` and `2.x-1.1`, as an example: `2.21.1` (version), `2.21.0`, `2.20.0`.

Reply:
The problem with semver comparison is that a pre-release or build-release is less than the final release (hence the word `Finalized` in the function. The traffic-manager version isn't final during development.